source/mainnet/smart-contracts/guides/unit-test-contract.rst

```diff
@@ -482,11 +482,169 @@ account:
+Property-based testing is supported using the |QuickCheck|_ crate.
+The tests should placed in the same module as Wasm unit tests and annotated with the ``#[concordium_quickcheck]`` macro.
+The macro takes a named attribute ``num_tests`` for specifying the number of random tests to run.
+If no ``num_tests`` given, the default number is ``100``.
+The return value of the function should be a boolean value with ``true`` corresponding to the fact that the property holds.
+In ``Cargo.toml``, ``concordium-std`` should be specified as a ``dev`` dependency with the ``concordium-quickcheck`` feature enabled.
+
+.. code-block::
+
+    ...
+
+    [dev-dependencies]
+    concordium-std = { version = "5", features = ["concordium-quickcheck"] }
+
+    ...
+
+In the code snipped below, the parameters ``address`` and ``amount`` are randomly generated.
+The process of generating random input and running the test is repeated ``num_tests = 500``.
+
+.. code-block:: rust
+
+    #[concordium_cfg_test]
+    mod test {
+
+       #[concordium_quickcheck(num_tests = 500)]
+       fn some_property_test(address: Address, amount: Amount) -> bool {
+        ...
+        // Instantiate custom struct with random parameters.
+        let input = MyParameters { sender: address, payment: amount }
+        ...
+        }
+    }
+
+The type like ``Address`` and ``Amount`` in the example have ``Arbitrary`` trait implementations, which are used to obtain random values.
+Read more about available ``Arbitrary`` instances for Concordium-specific type in |concordium_contracts_common|_ documentation.
+``Arbitrary`` instances for standard data types, like numbers and collections (``Vec``, ``BTreeMap``) are defined in |QuickCheck|_ and available automatically.
+Custom user data type instances can be created directly in tests using the random input parameters or by defining ``Arbitrary`` instances.
+See more details on QuickCheck's ``Arbitrary`` `here <https://docs.rs/quickcheck/latest/quickcheck/trait.Arbitrary.html>`_.
+
+The same command is used for running Wasm QuickCheck tests as in :ref:`tests_in_wasm`:
+
+.. code-block:: console
+
+    cargo concordium test
+
+When a test fails, it reports the random seed used to produce the input values.
+After making the required fixes one can use the same seed to see whether the tests works on the same generated values.
+The seed is a ``u64`` number, which can be provided along with the test command.
+
+.. code-block:: console
+
+    cargo concordium test --seed 1234567890
+
+Concordium QuickCheck tests can also be run with
+
+.. code-block:: console
+
+    cargo test
+
+By default, this command compiles the contract, unit tests and QuickCheck tests to machine code for your local target (most likely x86_64), and runs them.
+
+.. note::
+
+    Printing and supplying a seed is only possible using ``cargo concordium test``.
+
+.. warning::
+
+    Avoid using ``clam!`` and ``claim_eq!`` in ``#[concordium_quickcheck]`` tests.
```

Comment:
Perhaps something like:

```suggestion
    Avoid using ``claim!``, ``assert_eq!``, ``panic!`` or any other command that panics in ``#[concordium_quickcheck]`` tests.
```

Reply:
I think `clam!` and `claim_eq!` are "special": they don't just panic when running with `cargo-concordium`, they call `ReportError`. Using `assert` in Wasm would just give `Unreachable`.

I'll clarify this here as well. Because `claim!` will produce a user defined error message and interrupt the test. So, the counterexample won't be printed.

Reply:
Fixed now.